[BLOG] Amazon Q CLI Integration: Log Pattern and Data Distribution Analysis

PauiC opened the pull request:
### Description
Amazon Q CLI Integration: Log Pattern and Data Distribution Analysis 

### Issues Resolved
https://github.com/opensearch-project/project-website/issues/3985

### Check List
- [x] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

pajuric commented:
@xluo-aws - Adding you to this for peer review.

xluo-aws commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,199 @@
+date: 2025-10-17
+categories:
+  - technical-posts
+meta_keywords: Q CLI, CLI integration, log pattern analysis, data distribution analysis, OpenSearch CLI, command line interface, log analysis tools, data distribution tools, OpenSearch integration, diagnostic tools
```
Add MCP, Agent, Agentic AI, OpenTelemetry as keyword

xluo-aws commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,199 @@
+
+## Log Pattern Analysis Tool Overview
+
+The [Log Pattern Analysis Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/log-pattern-analysis-tool/) is an advanced OpenSearch agent tool that automates root cause analysis through multiple analysis modes. It performs differential pattern analysis between baseline and problem periods, analyzes log sequences using trace correlation, and provides log insights by autom ...
```
Advanced can be removed

PauiC commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,199 @@
+> 
+> 🛠️  Using tool: ListIndexTool (trusted) from mcp server opensearch
+>  ⋮ 
+>  ● Running ListIndexTool with the param:
```
This is a concrete example from my OpenTelemetry testing; I pasted the full input here. It does look incorrect, so I’ll delete these lines.

PauiC commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,199 @@
+
+## Q CLI Integration
+
+Q CLI can seamlessly integrate with these OpenSearch analysis tools through the [Model Context Protocol (MCP)](https://opensearch.org/blog/introducing-mcp-in-opensearch/). By configuring the OpenSearch MCP server, you can access both the Log Pattern Analysis Tool and Data Distribution Tool directly from your command line interface. This integration enables natural language queries for log analysi ...
```
Thanks for your suggestion. I've resolved it.

PauiC commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
````diff
@@ -0,0 +1,553 @@
+   ```
+
+2. **Add Tool Integration Code**
+   - Implement Log Pattern Analysis Tool integration in the MCP server
````
I forked the mcp-server repository and pushed my test code there. I’ve added the GitHub link in the blog.

xluo-aws commented:
@pajuric , we have completed internal review, could you help review as well?

kolchfa-aws commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,708 @@
+   - Other services: 21%
+
+2. **Severity level analysis** (divergence: 0.609):
+   - INFO level: 61% (normal operations)
```
It's not clear why there are different info levels. I'm assuming this is because the demo simulates different services where you have both java "INFO" and python "info" and maybe .net "Information"? If so, please add this in parentheses. If not, please add an explanatory sentence so readers can understand the discrepancy. Put actual error code levels in code font (so `error` or `INFO`).

PauiC replied:
Thanks for your suggestion. I've resolved it.

pajuric commented:
@PauiC - Please let me know once you have addressed @kolchfa-aws's comments and pushed them into the PR content.

PauiC commented:
> @PauiC - Please let me know once you have addressed @kolchfa-aws's comments and pushed them into the PR content.

@pajuric I’ve addressed all the comments and pushed the updates to the PR. Please let me know if you have any further feedback.

kolchfa-aws reviewed (dismissed):
Nice job, @PauiC, thanks! Moving on to editorial review.

natebower commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+
+## Data Distribution tool
+
+The [Data Distribution Tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/data-distribution-tool/) is an OpenSearch agent tool that analyzes data distribution patterns within datasets and compares distributions between different time periods. It supports both single dataset analysis and comparative analysis to identify significant changes in field value distributions, h ...
```
```suggestion
The [Data Distribution tool](https://docs.opensearch.org/latest/ml-commons-plugin/agents-tools/tools/data-distribution-tool/) is an OpenSearch agent tool that analyzes data distribution patterns within datasets and compares distributions between different time periods. It supports both single dataset analysis and comparative analysis to identify significant changes in field value distributions, helping detect anomalies, trends, and data quality issues.
```

natebower commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+
+### Amazon Q CLI investigation process
+
+Using Amazon Q CLI with the integrated OpenSearch tools, you can investigate this issue through natural language queries. Here's how the investigation unfolds:
```
```suggestion
Using Amazon Q CLI with the integrated OpenSearch tools, you can investigate this issue through natural language queries. Here's how the investigation unfolds.
```

natebower commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+
+**Core contribution: Automated error pattern recognition**
+
+The Log Pattern Analysis Tool played a crucial pattern identification role in this payment failure investigation:
```
```suggestion
The Log Pattern Analysis tool played a crucial pattern identification role in this payment failure investigation:
```

natebower commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+
+1. **Qualitative + quantitative analysis**: Log Pattern Analysis tool provided specific error patterns, Data Distribution tool provided statistical validation.
+2. **Priority guidance**: Combined analysis showed checkout service has disproportionately high failure impact despite lower log volume.
+3. **Root cause validation**: Both tools confirmed payment token validation as the primary issue, with product catalog as secondary.
```
```suggestion
3. **Root cause validation**: Both tools confirmed payment token validation as the primary issue, with the product catalog as secondary.
```

natebower commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+1. **Qualitative + quantitative analysis**: Log Pattern Analysis tool provided specific error patterns, Data Distribution tool provided statistical validation.
+2. **Priority guidance**: Combined analysis showed checkout service has disproportionately high failure impact despite lower log volume.
+3. **Root cause validation**: Both tools confirmed payment token validation as the primary issue, with product catalog as secondary.
+4. **Actionable insights**: Tools working together provided specific error messages and statistical significance, supporting clear remediation recommendations.
```
```suggestion
4. **Actionable insights**: The tools worked together to provide specific error messages and statistical significance, supporting clear remediation recommendations.
```

natebower reviewed (commented):
Editorial review

natebower reviewed (dismissed):
Thanks @PauiC! LGTM

@pajuric This should be ready to publish.

pajuric commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+authors:
+  - ihailong
+  - jiaruj
+date: 2025-10-17
```
Update: 2025-11-20

PauiC commented on _posts/2025-10-17-Q-CLI-Integration:-Log-Pattern-and-Data-Distribution-Analysis.md:
```diff
@@ -0,0 +1,709 @@
+categories:
+  - technical-posts
+meta_keywords: Amazon Q CLI, CLI integration, log pattern analysis, data distribution analysis, OpenSearch CLI, command line interface, log analysis tools, data distribution tools, OpenSearch integration, diagnostic tools, MCP, Agent, Agentic AI, OpenTelemetry
+meta_description: Learn how to integrate Amazon Q CLI with OpenSearch's log pattern analysis and data distribution tools for streamlined command-line diagnostics and enhanced system troubleshooting capabilities.
```
Thanks for your suggestion. I've resolved it.

natebower commented:
https://opensearch.org/blog/amazon-q-cli-integration-log-pattern-and-data-distribution-analysis/

natebower commented:
> @natebower - Please merge and close. Blog is live here: https://opensearch.org/blog/amazon-q-cli-integration-log-pattern-and-data-distribution-analysis/.

@pajuric The PR has already been merged.
